### ISSUE TITLE
[GSoC Discussion] Adding Random Walks

### ISSUE THREAD

**Issue by Smit-create** [sympy/sympy#19428]
A random walk is a stochastic process, that describes a path that consists of a succession of random steps on some mathematical space such as the integers.
A random walk is a Discrete-time discrete-state stochastic process similar to Bernoulli Process, but the state space in Integers. 
General API for Random Walk can be much similar to Bernoulli Process.
```python
>>> R = RandomWalk('R', p=0.3, reward=3, loss=-2) 
>>> E(R[3])
...
>>> P(R[3] > 2)   
...
>>> R.path(start_position=4, steps=10) # random generated path with given dist. of reward and loss
...
>>> expr = R[1] + R[2] + R[4] + R[3]  
>>> P(Eq(expr, 3))
...
>>> given_expr = R[1] + R[2]
>>> P(expr > 5, given_expr < 4)
...

**Comment by czgdp1807** [comment]
What about multi-dimensional RandomWalks? How can we represent the state in that case?

**Comment by Smit-create** [comment]
> How can we represent the state in that case?

We will by adding a CTSP, Wiener Process which is related to Random Walks, following (Normal Distribution)

**Comment by Smit-create** [comment]
For, the 1D Random Walks we can use Bernoulli Process, as Wiener Process is derived as limiting case of Random Walks

**Comment by Smit-create** [comment]
For, 2D Random walks, I am not getting strong notes, but from the Wikipedia photos of 2D and 3D, it seems that they considered independent one random walks in say x, y direction and then output in the form of x, y coordinates at a given time depending upon the starting position, reward and loss of x and y axes.

**Comment by czgdp1807** [comment]
Can you show how will RandomWalk API solve the problems and examples given in http://www2.math.uu.se/~sea/kurser/stokprocmn1/slumpvandring_eng.pdf ?
May be for multi-dimensional random walks we can try looking at other CAS? What do they have?

**Comment by Smit-create** [comment]
> http://www2.math.uu.se/~sea/kurser/stokprocmn1/slumpvandring_eng.pdf ?

I went through the problems, and one common similarity was the P(win) and P(loss), seeing their solution almost all the problems are solved using recurrence relation and doesn't use stats directly. I think we need to add two extra methods for this i.e. win and loss which solves using recurrence relation. 
It can be something like:
```python
>>> R = RandomWalk("R", p=1/2, reward=2, loss=-1)
>>> R[3] # this will sum of i.d.d BernoulliProcess rvs. i.e R[3] = B[0]+B[1]+B[2]+B[3]
...
>>> R.probability_win_at(7)
...
>>> R.probability_lose_at(-10)
...
```
Apart from these methods, other API can remain the same as:
```python
>>> P(Eq(R[6], 7))
...
>>> E(R[5] + R[2])
...
```

**Comment by czgdp1807** [comment]
> R.probability_win_at(7)

What do you mean by this?

**Comment by Smit-create** [comment]
> What do you mean by this?

It means what is the probability that the random walk will be at this point considering all the time-points.

**Comment by czgdp1807** [comment]
Can we do, `P(Eq(R[t], 7), Contains(t, R.index_set))`?

**Comment by Smit-create** [comment]
> Can we do, `P(Eq(R[t], 7), Contains(t, R.index_set))`?

I think we cannot because index_set is an infinite set, and also R[t] will be in unevaluated form

**Comment by Smit-create** [comment]
I think the only we can do is the following, also this can be time taking with large t as index, but will work quite good:
```ptython
>>> P(Eq(R[3], 7))
...
>>> P(R[4] > 5)
...

```

**Comment by czgdp1807** [comment]
How does the size of index set makes a difference? We are just trying to say that find the probability that the random walk will ever attain value 7. These kind of  situations can be handled by solving recurrences and sympy probably has some recurrence solvers. See, https://docs.sympy.org/latest/modules/solvers/solvers.html#sympy.solvers.recurr.rsolve
Actually, there can be queries like, find the probability whether a stochastic process attains a particular value sometime? It seems to be a pretty intuitive query and we should be capable of handling it.

**Comment by Smit-create** [comment]
> How does the size of index set makes a difference?

```python
>>> from sympy.stats import *
>>> from sympy import symbols, Sum
>>> B = BernoulliProcess('B', 0.5, 1, -1)
>>> t = symbols('t', positive=True, integer=True)
>>> expr = Sum(B[t], (t, 0, t)).doit() 
>>> P(expr > 7)   # does not work with infinite
...
RecursionError: maximum recursion depth exceeded while calling a Python object
>>> expr = Sum(B[t], (t, 0 , 10)).doit() # when t is finite, it works
>>> P(expr.doit() > 3)
0.113281250000000

```

> Actually, there can be queries like, find the probability whether a stochastic process attains a particular value sometime? It seems to be a pretty intuitive query and we should be capable of handling it.

I agree. But this needs to be handled separately as this will have to be provided with initial conditions to solve the recurrence, for example, in gambler's ruin recurrence in above pdf(page 6) needs `y(0)=0 `and `y(a+b)=1` to solve the recurrence, which is not obvious to be handled by P directly.

**Comment by czgdp1807** [comment]
We want to find the probability that the random walk is absorbed in state, `a+b` if it starts from state a. How do we frame such query using `RandomWalk` object and `P` function? One option can be, `P(Eq(R[t], a+b), Contains(t, R.index_set) & Eq(R[0], a))`. Now, how to detect from this query that the user is asking for absorption probability? We can use, the `sup` and `inf` of state space to check whether the `a+b` is an absorbing state. If it is an absorbing state, boundary conditions can be computed for the recurrence on the basis of whether that was `sup` or `inf`. Basically, we are trying to find probability of absorption of random walks in a particular state. And such queries can be handled by a separate private method of `RandomWalk` class. `t` is not of concern in such queries, it's just convey that we want chain to be absorbed at some point in time.

**Comment by Smit-create** [comment]
Another option can be, to pass the absorbing states as the time of initialization itself, i.e
`>>> R = RandomWalk("R", reward=+1, loss=-1, max_reward=a+b, max_loss=-b)`
Then to compute P for only values b/w -b and a+b elsewhere it will be zero. And if the P is asked to be computed at the extreme states, then use the recurrrence.

**Comment by czgdp1807** [comment]
> Then to compute P for only values b/w -b and a+b elsewhere it will be zero. And if the P is asked to be computed at the extreme states, then use the recurrrence.

We can do that as well but using sup and inf will be better. Suppose, I am giving the state_space as `{1, 2, 3, 4, 5, 6, 7, 8, 10}` then passing 1 and 10 again as absorbing state will be redundant. Can we use some more generic terms than `reward` and `loss`.

**Comment by Smit-create** [comment]
So this can be:
`>>> R = RandomWalk('R', p,  success=+1, fail=-1, state_space=Range(-10, 10), initial_position=0)`

For queries like '`will ever reach at point x`'  we will do the following:
1. y(initial_position)=0 and y(state_space._sup)=1
2. solve y(n) = p*y(n+success) + (1-p)*y(n+loss)
3. and output y(x)

**Comment by czgdp1807** [comment]
The API looks good. I think there should be a provision of giving initial position through probability queries as well.

**Comment by Upabjojr** [comment]
> `>>> R = RandomWalk('R', p, success=+1, fail=-1, state_space=Range(-10, 10), initial_position=0)`

I think `initial_position` is a too specific parameter. In general, I would expect the condition to be specified at any position. Maybe it should not be a parameter of `RandomWalk`?

**Comment by czgdp1807** [comment]
I agree too. Initial position can be passed as given condition in probability queries.

**Comment by Smit-create** [comment]
Agreed. It can be removed. But, still, I had a doubt if we could handle the queries of `P(Eq(X(t), 6), Contains(t, X.index_state)` separately.... than could prevent messing up with probability

**Comment by czgdp1807** [comment]
Have you thought of a private method inside class `RandomWalk` which can be called from inside the `RandomWalk.probability`. That logic would be isolated from the rest of stats module, so no conflicts with other things.
